_published/0013-authentication.md

```diff
@@ -0,0 +1,171 @@
+To keep the scope manageable the task of dividing up and refining access
+levels and permission will be left to a later date. For this RFC all kinds
+of read permissions won't require valid tokens, but all write type requests
+will require valid ID-tokens that have been signed by a pre-approved issuer.
+The OIDC protocol defines a token called an
+[Access Token](https://auth0.com/docs/security/tokens/access-tokens).
+The access token is signed by the source issuer. This signature is then used to
+validity and authenticity of the token. See
+[validating an OIDC Access Token](https://auth0.com/docs/security/tokens/access-tokens/validate-access-tokens).
+
+### Session Management
+
+The wharf-api is currently stateless and does not use sessions. It is
+possible to either include the token in every request to the server.
+
+### Local development vs production environments
+
+For the initial release of the feature the configuration will be tailored to
+using azure as an IDP. These settings will as necessary be exposed as
+environment variables to each respective docker image.
+
+For use in offline developer environments it should be possible to switch on a
+developer mode switch that allows bypassing validation of access
+tokens such that a valid login is not required.
+
+### Choice of Client
+
+The OIDC standard defines a
+[list of certified clients](https://openid.net/developers/certified/)
+that can help simplify the implementation of ID Token retrieval.
+
+For the backend a full client will not be required. As the tokens just need
+to be validated and not requested. See
+[validating an access token](https://auth0.com/docs/security/tokens/access-tokens/validate-access-tokens)
+
+## Compatibility
+
+Expect no compatability between versions of wharf without authentication to
+versions of wharf with authentication. There is a potential threat in leaving
+the system backwards compatible in that a non-secured version of the api could
+run together in an otherwise secure cluster.
+
+We should aim to pick a solution where any IDP supporting OIDC and RBAC can be
+used. Locking ourselves to a single IDP when a standard such as OIDC exists
+seems like a  waste opportunity. Keeping the choice of IDP open will leave
+more room for choice and flexibility in the future.
+
+## Alternative solutions
+
+There are many types of identity verification systems.
+This RFC assumes that OIDC will be used. Within OIDC there are multiple
+different [flow definitions](https://developer.okta.com/docs/concepts/oauth-openid/#choosing-an-oauth-2-0-flow).
+We should primarily use `codeflow with pkce` as this is presumed to be the
+most secure at the time of writing.
+
+## Future possibilities
+
+We may want to implement multiple different IDP sources. This could allow
+login from different IDPs. A sample situation where this could be useful is
+when a third party view/control over the deployment pipeline.
```

Comment:
Some words seem to be missing here.

Reply:
Now - `A sample situation where this could be useful is
when an external third party should be granted view or admin rights over the
deployment pipeline.`